zio-http/src/main/scala/zio/http/Form.scala

```diff
@@ -0,0 +1,71 @@
+package zio.http
+
+import java.net.URLEncoder
+import java.nio.charset.Charset
+import java.net.URLDecoder
+import zio._
+
+/**
+ * Represents a form that can be either multipart or url encoded.
+ */
+sealed trait Form { self =>
+
+  def +(field: (String, String)): Form = appendField(field._1, field._2)
+
+  def appendField(name: String, value: String): Form =
+    self match {
+      case Form.URLEncodedForm(encoding, fields) =>
+        Form.URLEncodedForm(encoding, fields + (name -> value))
+    }
+
+  def contentType = self match {
+    case Form.URLEncodedForm(_, _) => "application/x-www-form-urlencoded"
+  }
+
+  def fields: Map[String, String]
+
+}
+
+object Form {
+
+  private val `UTF-8` = Charset.forName("UTF-8")
+
+  /**
+   * Represents a `application/x-www-form-urlencoded` form.
+   *
+   * `Request.form` will automatically detect and set the `Content-Type` header
+   * to `application/x-www-form-urlencoded` for this request.
+   *
+   * @param value
+   */
+  private[zio] final case class URLEncodedForm(private[zio] encoding: Charset, fields: Map[String, String])
+      extends Form {
+    def toEncodedString: String = encodeFields(encoding, fields).map { case (k, v) => s"$k=$v" }.mkString("&")
+  }
+
+  object URLEncodedForm {
+    def decode(encoded: String, encoding: Charset): Task[Form] = {
```

Comment:
This should be pure as it does not perform side-effects. Also, it should be called `From.fromDecoded`, and we can delete the `object URLEncodedForm`.